File: week14/20/pr3_Next_Greater_Element_I.py
from typing import List
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        ans = []
        n = len(nums2)
        for x in nums1:
            idx = nums2.index(x)
            cond = False
            for y in range(idx+1,n):
                if nums2[y] > x :
                    ans.append(nums2[y])
                    cond = True
                    break
            if not cond : ans.append(-1)
        
        return ans
    
# class Solution:
#     def nextGreaterElement(self, nums1, nums2):
#         stack = []
#         hashmap = {}

#         for num in nums2:
#             while stack and num > stack[-1]:
#                 hashmap[stack.pop()] = num
#             stack.append(num)

#         # while stack:
#         #     hashmap[stack.pop()] = -1

#         return [hashmap.get(i, -1) for i in nums1] #-1 is default value if the key i isn't present in the hashmap

File: week14/20/pr3_Next_Greater_Element_I.cpp
using namespace std;
#include "vector"
#include <stack>
#include <map>

class Solution {
public:
    vector<int> nextGreaterElement(vector<int>& nums1, vector<int>& nums2) {
        stack<int> s;
        map<int,int> m;
        
        for(int i=nums2.size()-1;i>=0;i--){
            int temp = nums2[i];
            while(!s.empty() && temp > s.top()){
                s.pop();
            }
            if(s.empty()) m[temp] = -1;
            else m[temp] = s.top();
            s.push(temp);
        }
        for(int j=0;j<nums1.size();j++){
            nums1[j] = m[nums1[j]];
        }
        return nums1;
    }
};